File: advent_of_code/2015/1/solution.py
class Solver():
    def __init__(self):
        self.input = open('input.txt', 'r').read()
        self.floor = 0
        # for solve_2, to track the position of the char
        self.current_char = 0
        self.basement_logged = False

    def move(self):
        for bracket in self.input:
            if self.floor == -1 and not self.basement_logged:
                print(f'Santa enters the basement at {self.current_char}')
                self.basement_logged = True
            if bracket == '(':
                self.floor +=1
            elif bracket == ')':
                self.floor -= 1
            self.current_char += 1

    def reset_counters(self):
        self.current_char = 0
        self.floor = 0

    def solve(self):
        self.move()
        print(f'Santa ended on floor {self.floor} in part 1')
        self.reset_counters()

solver = Solver()
solver.solve()

File: advent_of_code/2015/1/solution.js
/*
 * `(` - open bracket symbol, takes santa one floor up.
 * `)` - closed bracket symbol, takes him one floor down.
 *
 * given a file containing those parens as input, answer: which floor
 * does santa get to?
 *
 */

const fs = require("fs");

const solveSantasProblem = () => {
  return new Promise((resolve, reject) => {
    const processFloorInput = (err, floorText) => {
      if (err) {
        reject(err);
      }

      let floor = 0;
      let visitedBasement = false;
      let indexOfBasementVisitCharacter;

      for (let i = 0; i < floorText.length; i++) {
        const char = floorText[i];
        console.log(i, char, floor);
        if (!visitedBasement) {
          if (floor === -1) {
            visitedBasement = true;
            indexOfBasementVisitCharacter = i;
          }
        }

        switch (char) {
          case "(":
            floor++;
            break;
          case ")":
            floor--;
            break;
          default:
            break;
        }
      }
      resolve({ floor, indexOfBasementVisitCharacter });
    };
    fs.readFile("./input.txt", "utf8", processFloorInput);
  });
};

const solve = async () => {
  const { floor, indexOfBasementVisitCharacter } = await solveSantasProblem();
  console.log(
    `Santa will end up on floor ${floor}. First time in basement on character ${indexOfBasementVisitCharacter}`
  );
};

solve();
